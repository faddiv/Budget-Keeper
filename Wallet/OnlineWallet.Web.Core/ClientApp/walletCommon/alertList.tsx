import * as React from "react";
import { bindActionCreators, compose } from "redux";
import { RouteComponentProps, withRouter } from "react-router";
import { connect } from "react-redux";
import { bind } from "bind-decorator";

import { AlertsActions } from "actions/alerts";
import { AlertMessage } from "reducers/alerts/alertsModel";
import { RootState } from "reducers";

export interface AlertListProps extends Partial<RouteComponentProps<void>> {
    alerts?: AlertMessage[];
    actions?: typeof AlertsActions;
}

export interface AlertListState {
}

export class AlertList2 extends React.Component<AlertListProps, AlertListState> {

    private navListener: () => void;

    constructor(props) {
        super(props);
        this.state = {
            items: []
        };
    }

    @bind
    private locationListener() {
        const { alerts, actions } = this.props;
        if (alerts.length > 0) {
            actions.dismissAllAlert();
        }
    }

    componentDidMount() {
        this.navListener = this.props.history.listen(this.locationListener);
    }

    componentWillUnmount() {
        this.navListener();
        this.navListener = null;
    }

    dismissAlert(alert: AlertMessage) {
        this.props.actions.dismissAlert(alert);
    }

    @bind
    renderAlert(alert: AlertMessage, index: number) {
        const css = "alert alert-" + alert.type;
        return (
            <div key={index} className={css} role="alert">
                {alert.message}
                <button type="button" className="close" aria-label="Close" onClick={() => this.dismissAlert(alert)}>
                    <span aria-hidden="true">&times;</span>
                </button>
            </div>
        );
    }

    render() {
        return (
            <div className="alerts">
                {this.props.alerts.map(this.renderAlert)}
            </div>
        );
    }
}

function mapStateToProps(state: RootState) {
    return {
        alerts: state.alerts
    };
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(AlertsActions as any, dispatch) as typeof AlertsActions
    };
}

export const AlertList = compose(withRouter, connect(mapStateToProps, mapDispatchToProps))(AlertList2);
