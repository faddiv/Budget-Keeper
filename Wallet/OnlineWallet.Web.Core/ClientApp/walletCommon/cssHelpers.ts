export function className(...values: (string | boolean)[]): string {
    let show = true;
    let cssList: string[] = [];
    for (var value of values) {
        if (typeof (value) !== "string") {
            show = !!value;
            continue;
        }
        if (!show) {
            continue;
        }
        cssList.push(value);
    }
    if (cssList.length) {
        return cssList.join(" ");
    }
}