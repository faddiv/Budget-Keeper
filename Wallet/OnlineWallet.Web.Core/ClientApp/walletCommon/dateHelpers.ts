import * as moment from "moment";

export function toUTCDate(value: any): Date | undefined {
    if (value) {
        var date = moment(value);
        return new Date(Date.UTC(date.year(), date.month(), date.date()));
    }
}