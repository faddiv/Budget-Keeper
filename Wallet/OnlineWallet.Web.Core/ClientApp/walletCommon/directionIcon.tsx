import * as React from "react";
import { MoneyDirection } from "walletApi";

interface DirectionIconProps {
    direction: MoneyDirection;
}

export const DirectionIcon: React.SFC<DirectionIconProps> = ({ direction, ...rest }) => {
    return (
        <span className={directionCssClass(direction)}></span>
    );
};

function directionCssClass(direction: MoneyDirection) {
    switch (direction) {
        case -1:
            return "fa fa-minus text-danger";
        case 1:
            return "fa fa-plus text-success";
        default:
            return "fa fa-bookmark";
    }
}
