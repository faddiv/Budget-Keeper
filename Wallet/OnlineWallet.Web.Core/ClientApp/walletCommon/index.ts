export * from "./alertList";
export * from "./directionIcon";
export * from "./editDelete";
export * from "./formGroup";
export * from "./saveCancel";
export * from "./pager";
export * from "./navbar";
export * from "./models";
export * from "./specialInputs";
export * from "./transactions-view";