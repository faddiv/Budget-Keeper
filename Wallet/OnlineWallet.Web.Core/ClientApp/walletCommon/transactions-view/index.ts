export * from "./transaction-table/transaction-table";