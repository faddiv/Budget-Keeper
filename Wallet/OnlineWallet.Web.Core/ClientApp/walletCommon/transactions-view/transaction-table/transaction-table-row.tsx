import * as React from "react";
import * as classNames from "classnames";
import { bind } from "bind-decorator";

import { Wallet, ArticleModel, CategoryModel } from "walletApi";
import { DirectionIcon, SaveCancel, EditDelete, ITransactionTableExtFunction, WalletSelector, NameInput, CategoryInput, TransactionViewModel, nextDirection, getWalletNameById } from "walletCommon";
import { noop } from "helpers";
import { updateState, isClickableClicked } from "react-ext";

export interface TransactionTableRowProps {
    item: TransactionViewModel;
    selected: boolean;
    wallets: Wallet[];
    rowColor?: ITransactionTableExtFunction;
    editable: boolean;
    saveTransaction(newItem: TransactionViewModel, originalItem: TransactionViewModel): void;
    deleteTransaction(item: TransactionViewModel): void;
    rowMouseDown(item: TransactionViewModel): void;
    rowMouseUp(): void;
    rowMouseEnter(item: TransactionViewModel): void;
}

export interface TransactionTableRowState {
    editMode: boolean;
    item: TransactionViewModel;
    original: TransactionViewModel;
}

export class TransactionTableRow extends React.Component<TransactionTableRowProps, TransactionTableRowState> {

    constructor(props: TransactionTableRowProps) {
        super(props);
        this.state = {
            editMode: false,
            item: props.item,
            original: props.item
        };
    }

    componentWillReceiveProps(nextProps: Readonly<TransactionTableRowProps>) {
        if (this.state.original !== nextProps.item) {
            this.setState({
                item: nextProps.item,
                original: nextProps.item
            });
        }
    }

    @bind
    changeDirection(event: React.MouseEvent<HTMLTableDataCellElement>) {
        event.preventDefault();
        this.setState((prevState) => {
            return {
                item: { ...prevState.item, direction: nextDirection(prevState.item.direction) }
            };
        });
    }

    @bind
    saveTransaction() {
        this.props.saveTransaction(this.state.item, this.props.item);
        this.setState({
            editMode: false
        });
    }

    @bind
    deleteTransaction() {
        this.props.deleteTransaction(this.props.item);
    }

    @bind
    editTransaction() {
        this.setState({
            editMode: true
        });
    }

    @bind
    cancelTransaction() {
        this.setState((_prevState, props) => ({
            editMode: false,
            item: props.item
        }));
    }

    @bind
    handleInputChange(event: React.SyntheticEvent<HTMLElement>) {
        const state = updateState(event);
        if (state.walletId) {
            state.walletId = parseInt(state.walletId, 10);
        }
        this.setState((prevState) => {
            return {
                item: { ...prevState.item, ...state }
            };
        });
    }

    @bind
    nameSelected(model: ArticleModel) {
        this.setState((prevState) => {
            return {
                item: {
                    ...prevState.item,
                    name: model.name
                }
            };
        });
    }

    @bind
    categorySelected(model: CategoryModel) {
        this.setState((prevState) => {
            return {
                item: {
                    ...prevState.item,
                    category: model.name
                }
            };
        });
    }

    @bind
    startSelection(event: React.MouseEvent<HTMLTableRowElement>) {
        if (event.isDefaultPrevented() ||
            isClickableClicked(event)) {
            return;
        }
        this.props.rowMouseDown(this.props.item);
        event.preventDefault();
    }

    @bind
    continueSelection(event: React.MouseEvent<HTMLTableRowElement>) {
        if (event.isDefaultPrevented() ||
            isClickableClicked(event)) {
            return;
        }
        this.props.rowMouseEnter(this.props.item);
    }

    @bind
    endSelection(event: React.MouseEvent<HTMLTableRowElement>) {
        if (event.isDefaultPrevented() ||
            isClickableClicked(event)) {
            return;
        }
        this.props.rowMouseUp();
    }

    render() {
        return this.state.editMode ? this.renderEditRow() : this.renderViewRow();
    }

    renderEditRow() {
        const { item } = this.state;
        const { wallets, rowColor, selected } = this.props;
        const hasRowColor = !!rowColor;
        return (
            <tr className={classNames({ [rowColor(item)]: hasRowColor }, { selected })} onChange={this.handleInputChange}
                onMouseDown={this.startSelection} onMouseEnter={this.continueSelection} onMouseUp={this.endSelection}>
                <td>
                    <input type="date" className="form-control" value={item.createdAt} name="createdAt" onChange={noop} />
                </td>
                <td>
                    <NameInput value={item.name} className="form-control" onSelect={this.nameSelected} />
                </td>
                <td onClick={this.changeDirection}>
                    <DirectionIcon direction={item.direction} />
                </td>
                <td>
                    <input type="number" className="form-control" value={item.price} name="price" autoComplete="off" onChange={noop} />
                </td>
                <td>
                    <WalletSelector walletId={item.walletId} wallets={wallets} className="form-control" />
                </td>
                <td>
                    <CategoryInput value={item.category} onSelect={this.categorySelected} className="form-control" />
                </td>
                <td>
                    <input type="text" className="form-control" value={item.comment} name="comment" autoComplete="off" onChange={noop} />
                </td>
                <td>
                    <SaveCancel save={this.saveTransaction} cancel={this.cancelTransaction} />
                </td>
            </tr>
        );
    }

    renderViewRow() {
        const { rowColor, wallets, selected, editable } = this.props;
        const { item } = this.state;
        const hasRowColor = !!rowColor;
        return (
            <tr className={classNames({ [rowColor(item)]: hasRowColor }, { selected })}
                onMouseDown={this.startSelection} onMouseEnter={this.continueSelection} onMouseUp={this.endSelection}>
                <td>{item.createdAt}</td>
                <td>{item.name}</td>
                <td>
                    <DirectionIcon direction={item.direction} />
                </td>
                <td>{item.price}</td>
                <td>{getWalletNameById(item.walletId, wallets)}</td>
                <td>{item.category}</td>
                <td>{item.comment}</td>
                {editable &&
                    <td>
                        <EditDelete edit={this.editTransaction} delete_={this.deleteTransaction} />
                    </td>}
            </tr>
        );
    }
}
