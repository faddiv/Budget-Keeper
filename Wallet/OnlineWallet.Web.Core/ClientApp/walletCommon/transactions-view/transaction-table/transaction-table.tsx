import * as React from "react";
import { Wallet } from "walletApi";
import { TransactionTableRow } from "./transaction-table-row";
import { TransactionViewModel, ITransactionTableExtFunction } from "walletCommon";
import { bind, _ } from "helpers";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { TransactionSummaryActions } from "actions/transactionsSummary";
import { RootState } from "reducers";

enum SelectMode {
    deselect,
    none,
    select
}

export interface TransactionTableProps {
    actions?: typeof TransactionSummaryActions;
    items: TransactionViewModel[];
    wallets: Wallet[];
    changedItems?: TransactionViewModel[];
    rowColor?: ITransactionTableExtFunction;
    transactionSummary?: TransactionViewModel[];
    update?(items: TransactionViewModel[], changedItems?: TransactionViewModel[]): void;
    deleted?(items: TransactionViewModel): void;
}

export interface TransactionTableState {
    selectMode: SelectMode;
}

@connect(mapStateToProps, mapDispatchToProps)
export class TransactionTable extends React.Component<TransactionTableProps, TransactionTableState> {

    constructor(props: TransactionTableProps) {
        super(props);
        this.state = {
            selectMode: SelectMode.none
        };
    }

    @bind
    saveTransaction(newItem: TransactionViewModel, original: TransactionViewModel) {
        const items = _.replace(this.props.items, newItem, original);
        const changes = this.props.changedItems ? _.replace(this.props.changedItems, newItem, original, true) : undefined;
        this.props.update(items, changes);
    }

    @bind
    deleteTransaction(item: TransactionViewModel) {
        this.props.deleted(item);
    }

    @bind
    startSelection(item: TransactionViewModel) {
        this.setState((_prevState: TransactionTableState, props: TransactionTableProps) => {
            const selectMode = _.contains(props.transactionSummary, item)
                ? SelectMode.deselect
                : SelectMode.select;
            const selected = selectMode === SelectMode.select
                ? [...props.transactionSummary, item]
                : _.remove(props.transactionSummary, item);
            props.actions.transactionsSelected(selected);
            return {
                selectMode,
                selected
            };
        });
    }

    @bind
    selectingRow(item: TransactionViewModel) {
        if (this.state.selectMode === SelectMode.none) { return; }
        let selected: TransactionViewModel[];
        if (this.state.selectMode === SelectMode.select) {
            if (!_.contains(this.props.transactionSummary, item)) {
                selected = [...this.props.transactionSummary, item];
            }
        } else {
            if (_.contains(this.props.transactionSummary, item)) {
                selected = _.remove(this.props.transactionSummary, item);
            }
        }
        if (selected) {
            this.props.actions.transactionsSelected(selected);
        }
    }

    @bind
    endSelection() {
        this.setState({
            selectMode: SelectMode.none
        });
    }

    render() {
        const { items, wallets, rowColor, transactionSummary, update } = this.props;
        const editable = !!update;
        return (
            <table className="table transactions" onMouseLeave={this.endSelection}>
                <thead>
                    <tr>
                        <th className="created-at">createdAt</th>
                        <th className="name">name</th>
                        <th className="direction">dir</th>
                        <th className="price">price</th>
                        <th className="wallet-name">walletName</th>
                        <th className="category">category</th>
                        <th>comment</th>
                        {editable && <th className="commands"></th>}
                    </tr>
                </thead>
                <tbody>
                    {items.map(item =>
                        <TransactionTableRow
                            key={item.key}
                            item={item}
                            selected={_.contains(transactionSummary, item)}
                            wallets={wallets}
                            deleteTransaction={this.deleteTransaction}
                            saveTransaction={this.saveTransaction}
                            editable={editable}
                            rowColor={rowColor}
                            rowMouseDown={this.startSelection}
                            rowMouseEnter={this.selectingRow}
                            rowMouseUp={this.endSelection} />)}
                </tbody>
            </table>
        );
    }
}

function mapStateToProps(state: RootState) {
    return {
        transactionSummary: state.transactionSummary
    };
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(TransactionSummaryActions as any, dispatch) as typeof TransactionSummaryActions
    };
}
