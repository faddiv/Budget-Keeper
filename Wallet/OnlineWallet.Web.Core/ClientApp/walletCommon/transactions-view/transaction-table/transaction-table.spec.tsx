import * as React from "react";
import { render, mount } from "enzyme";
import "jest-enzyme";

import { TransactionTable } from "./transaction-table";
import { unwrap } from "helpers/testHelpers";
import { TransactionSummaryActions } from "actions/transactionsSummary";
import { AlertsActions } from "actions/alerts";
import { createAlertsActionsMock } from "actions/alerts.mocks";
import { createTransactionSummaryActionsMocks } from "actions/transactionsSummary.mocks";
import { Wallet } from "walletApi";
import { TransactionViewModel, getDirectionColoring } from "walletCommon";

describe("TransactionSummary", () => {
    const TransactionTable2 = unwrap(TransactionTable);
    let summaryActions: typeof TransactionSummaryActions;
    let alertActions: typeof AlertsActions;
    const wallets: Wallet[] = [
        {
            moneyWalletId: 1,
            name: "Cash"
        }, {
            moneyWalletId: 2,
            name: "BankAccount"
        }
    ];

    beforeEach(() => {
        summaryActions = createTransactionSummaryActionsMocks();
        alertActions = createAlertsActionsMock();
    });

    it("should display transaction rows", () => {
        const items: TransactionViewModel[] = createTransactions(4);
        const element = render(<TransactionTable2 alertActions={alertActions}
            summaryActions={summaryActions} items={items} wallets={wallets}
            transactionSummary={[]} rowColor={getDirectionColoring} />);

        expect(element).toMatchSnapshot();
    });

    it("should highlight selected in summary rows", () => {
        const items: TransactionViewModel[] = createTransactions(4);
        const transactionSummary = items.slice(1, 3);

        const element = mount(<TransactionTable2 alertActions={alertActions}
            summaryActions={summaryActions} items={items} wallets={wallets}
            transactionSummary={transactionSummary} rowColor={getDirectionColoring} />);

        const rows = element.find("tbody tr");
        expect(rows.length).toBe(4);
        expect(rows.at(0)).not.toHaveClassName("selected");
        expect(rows.at(1)).toHaveClassName("selected");
        expect(rows.at(2)).toHaveClassName("selected");
        expect(rows.at(3)).not.toHaveClassName("selected");
    });
});

function createTransactions(size: number = 4) {
    return [
        { key: 1, transactionId: 1, name: "apple", category: "grocery", comment: "Hy", createdAt: "2018-05-26", price: "123", direction: -1, walletId: 1 },
        { key: 2, transactionId: 2, name: "payment", category: "incomes", comment: "Got it", createdAt: "2018-05-27", price: "30000", direction: 1, walletId: 2 },
        { key: 3, transactionId: 3, name: "detergent", category: "cleaning", comment: "for cleaning", createdAt: "2018-05-27", price: "100", direction: -1, walletId: 1 },
        { key: 4, transactionId: 4, name: "movie ticket", category: "fun", comment: "good movie", createdAt: "2018-05-28", price: "1000", direction: 0, walletId: 2 }
    ].slice(0, size);
}
