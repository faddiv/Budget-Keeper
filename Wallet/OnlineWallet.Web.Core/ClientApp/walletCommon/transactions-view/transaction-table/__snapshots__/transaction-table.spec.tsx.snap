// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionSummary should display transaction rows 1`] = `
<table
  class="table transactions"
>
  <thead>
    <tr>
      <th
        class="created-at"
      >
        createdAt
      </th>
      <th
        class="name"
      >
        name
      </th>
      <th
        class="direction"
      >
        dir
      </th>
      <th
        class="price"
      >
        price
      </th>
      <th
        class="wallet-name"
      >
        walletName
      </th>
      <th
        class="category"
      >
        category
      </th>
      <th>
        comment
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      class="undefined"
    >
      <td>
        2018-05-26
      </td>
      <td>
        apple
      </td>
      <td>
        <span
          class="fa fa-minus text-danger"
        />
      </td>
      <td>
        123
      </td>
      <td>
        Cash
      </td>
      <td>
        grocery
      </td>
      <td>
        Hy
      </td>
    </tr>
    <tr
      class="table-success"
    >
      <td>
        2018-05-27
      </td>
      <td>
        payment
      </td>
      <td>
        <span
          class="fa fa-plus text-success"
        />
      </td>
      <td>
        30000
      </td>
      <td>
        BankAccount
      </td>
      <td>
        incomes
      </td>
      <td>
        Got it
      </td>
    </tr>
    <tr
      class="undefined"
    >
      <td>
        2018-05-27
      </td>
      <td>
        detergent
      </td>
      <td>
        <span
          class="fa fa-minus text-danger"
        />
      </td>
      <td>
        100
      </td>
      <td>
        Cash
      </td>
      <td>
        cleaning
      </td>
      <td>
        for cleaning
      </td>
    </tr>
    <tr
      class="table-warning"
    >
      <td>
        2018-05-28
      </td>
      <td>
        movie ticket
      </td>
      <td>
        <span
          class="fa fa-bookmark"
        />
      </td>
      <td>
        1000
      </td>
      <td>
        BankAccount
      </td>
      <td>
        fun
      </td>
      <td>
        good movie
      </td>
    </tr>
  </tbody>
</table>
`;
